### PR TITLE
move batch to device before sending it to hooks

### LINKED ISSUE #6990: Eliminate redundant device moves for callbacks
## 🚀 Feature

Make it possible to access batches on callbacks without moving them to the CPU

### Motivation

Currently, on callback hooks like `on_train_batch_end` where a `batch` tensor is available, the tensor is on cpu by default. If I want to, for example, run that batch through another model, I have to move it back to CUDA with `batch.to('cuda')` even though it must have been on CUDA in order for the training step to have taken place. This creates a serious performance bottleneck.

### Pitch

Either add an argument to `on_train_batch_end` and similar hooks for cuda or add new hooks for before the move takes place.

### Additional context

https://github.com/PyTorchLightning/pytorch-lightning/discussions/6945

### LINKED ISSUE #7377: on_train_batch_end receives batch on CPU
## 🐛 Bug

The callback method on_train_batch_end receives the batch still on CPU instead of the batch that the accelerator moved to the device.

### To Reproduce

https://colab.research.google.com/drive/1axSx8lmpAhllnlcga6iE2eNgC1fjGK1b?usp=sharing

### Expected behavior

The `on_train_batch_end` hook receives the batch on the device. 

### Environment

 CUDA:
        - GPU:
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
                - GeForce RTX 3090
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1+cu111
        - pytorch-lightning: 1.3.0rc2
        - tqdm:              4.60.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.8
        - version:           #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020

### Additional context

Asked in the discussion:

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #7377
Fixes #6990

In the loops we have the following structure:

```python
on_train_batch_start(batch)  # master: batch on CPU, this PR: batch is on GPU
accelerator.training_step(batch) # master: batch gets moved to GPU here, this PR: batch is already on GPU here
on_train_batch_end(batch)  # master: batch on CPU, this PR: batch is on GPU

```

**Test fails on master.**

**Open to suggestions for a more elegant way to test.**

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -954,6 +954,9 @@ def run_evaluation(self, on_epoch: bool = False) -> _EVALUATE_OUTPUT:
                 if batch_idx >= dl_max_batches:
                     break
 
+                # TODO: where is the profile step for to_device?
+                batch = self.accelerator.to_device(batch)
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:959]: does the same apply for the predict loop?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:959]: Yes, will continue the PR shortly. Not sure if this will be correct, havent checked in detail.

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: done

**Review by ananthsub — APPROVED** [review]
This is much more explicit!

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
LGTM

**Comment by awaelchli** [comment]
Hmm, benchmark tests show that now we are using more memory than vanilla PyTorch. There must be a copy somewhere that we end up with the batch twice in memory. Not sure where yet

**Review by s-rog — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
wow ! Really neat !

File: tests/trainer/loops/test_all.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+# Copyright The PyTorch Lightning team.
```
Comment (ananthsub) [inline tests/trainer/loops/test_all.py:1]: do you prefer one catch-all test? or dedicated test files for each loop?

**Comment by awaelchli** [comment]
@rohitgr7 do you have advice how to resolve this PR after #6241? `dataloader_idx` is `None` instead of 0, supposedly on purpose when using a single dataloader.

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -383,13 +375,6 @@ def setup_precision_plugin(self) -> None:
         self.optimizers = optimizers
         self.schedulers = schedulers
 
-    def to_device(self, step_kwargs: Dict[str, Union[Any, int]]) -> Dict[str, Union[Any, int]]:
-        """Pushes the batch to the root device"""
-        step_kwargs['batch'] = self.batch_to_device(
-            step_kwargs['batch'], self.root_device, dataloader_idx=step_kwargs.get('dataloader_idx', None)
-        )
-        return step_kwargs
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/accelerator.py:391]: just fyi:
there's a to_device method in `accelerators/gpu.py`, you might need to take care of that too and update accordingly.

**Review by rohitgr7 — COMMENTED** [review]
regarding the dataloader_idx being None or 0 in the case of a single dataloader, I didn't give it much thought at that time since it's meaningless in such a case. 0 or None both seems reasonable to me.

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:391]: thanks, didn't see that.

**Review by justusschock — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by import-antigravity** [comment]
Any word on when this might be done?

File: tests/trainer/loops/test_all.py (hunk 2 above)
Comment (awaelchli) [reply tests/trainer/loops/test_all.py:1]: no strong preference, just following the pattern we have in other places too.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -174,7 +175,9 @@ def batch_to_device(
         """
         model = self.lightning_module
 
-        if model is not None:
+        # TODO: Add support to allow batch transfer to device in Lightning for DP mode.
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:178]: Already have this one

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/configuration_validator.py#L111

File: tests/trainer/loops/test_all.py (hunk 2 above)
Comment (carmocca) [reply tests/trainer/loops/test_all.py:1]: IMO this is okay 👍

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -504,6 +508,9 @@ def run_training_epoch(self):
                 dataloader_idx,
             )
 
+            # clear memory before running validation
+            del batch
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: Well good thing we added it otherwise we wouldn't have noticed.

But I don't understand why it's required now. Does the garbage collector not collect CUDA garbage?

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: it gets marked for garbage collection when the scope of the local variable ends. 
so before, this batch was moved to the device inside the hook call which limited the memory use to that scope. now that I have moved it here, the local variable batch will stay in memory as long as we are in the scope of this training for loop and since we are also running validation here, there is an overlap. 

good news is, in our loop refactor we can again remove the del batch because we will be able to separate the training step from the validation call

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: that being said, GPU parity benchmark is still not happy comparing to PyTorch so I can dig even more this weekend ... :( haha

[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/trainer/training_loop.py]: @carmocca Per default it doesn't collect CUDA-garbage. It marks the Tensor object (which is only a pointer to CUDA memory AFAIK) for deletion and PyTorch has it's own garbage collector for CUDA based on this I think.

File: benchmarks/test_basic_parity.py [hunk 6]
```diff
@@ -45,8 +45,8 @@ def assert_parity_absolute(pl_values, pt_values, norm_by: float = 1, max_diff: f
 @pytest.mark.parametrize(
     'cls_model,max_diff_speed,max_diff_memory',
     [
-        (ParityModuleRNN, 0.05, 0.0),
-        (ParityModuleMNIST, 0.25, 0.0),  # todo: lower this thr
+        (ParityModuleRNN, 0.05, 0.001),
+        (ParityModuleMNIST, 0.25, 0.001),  # todo: lower this thr
```
Comment (awaelchli) [inline benchmarks/test_basic_parity.py:49]: @PyTorchLightning/core-contributors  Am I allowed to to this? 

the memory measured here in this benchmark is `torch.cuda.max_memory_allocated()`, the max cuda memory allocated during the process. This branch uses a bit more memory because when we move the batch to the device now in the new location, the batch remains in memory a bit longer than before (because we need to keep it until we call on_train_batch_end hook) and it overlaps with other memory being used by the loop, so the max has slightly increased, going over the benchmark threshold here. Let me know what you think.

**Review by awaelchli — COMMENTED** [review]
Re-requesting review from all. Since the creation of this PR we moved to the new loop API.

File: pytorch_lightning/accelerators/gpu.py [hunk 7]
```diff
@@ -51,11 +49,3 @@ def set_nvidia_flags(local_rank: int) -> None:
         all_gpu_ids = ",".join([str(x) for x in range(torch.cuda.device_count())])
         devices = os.getenv("CUDA_VISIBLE_DEVICES", all_gpu_ids)
         _log.info(f"LOCAL_RANK: {local_rank} - CUDA_VISIBLE_DEVICES: [{devices}]")
-
-    def to_device(self, step_kwargs: Dict[str, Union[Any, int]]) -> Dict[str, Union[Any, int]]:
-        # no need to transfer batch to device in DP mode
-        # TODO: Add support to allow batch transfer to device in Lightning for DP mode.
```
Comment (carmocca) [inline pytorch_lightning/accelerators/gpu.py:57]: Is this limitation removed?

Do you need to also update `ConfigValidator.__verify_dp_batch_transfer_support`?

**Review by carmocca — APPROVED** [review]

File: benchmarks/test_basic_parity.py (hunk 6 above)
Comment (carmocca) [reply benchmarks/test_basic_parity.py:49]: The memory difference would be back to 0 if the batch was also kept alive up until the same point in the native version, right? In that case, then it's okay with me.

Also, is 0.001 = 0.1%?

File: pytorch_lightning/accelerators/gpu.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/gpu.py:57]: no, it is not. 

I have moved this comment and the check to the accelerator class where we handle the batch transfer. 
I removed the TODO because it is so extremely unrealistic that we can ever add support for this, unless pytorch adds some hooks to DataParallel that we can use for a custom scatter and gather.

File: benchmarks/test_basic_parity.py (hunk 6 above)
Comment (awaelchli) [reply benchmarks/test_basic_parity.py:49]: Yes, if the vanilla loop also did all the stuff between here https://github.com/PyTorchLightning/pytorch-lightning/blob/af52de1198e9e17dcea57b4e29ad173e044fd85b/pytorch_lightning/loops/epoch/training_epoch_loop.py#L118 and L147 that we do in lightning. 

> Also, is 0.001 = 0.1%?

Yes this is a relative memory difference, measured like so:

`rel_diff = (pl_value - pt_value) / pt_value`

we then check:
`rel_diff < 0.001 = 0.1%`

Comment (ananthsub) [reply benchmarks/test_basic_parity.py:49]: IMO this is acceptable such that we have consistency on where the batch is across module steps and callback hooks. Also this is the first time I've seen the parity trigger for memory like this. Very cool it works like that at this sensitivity!

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -173,8 +174,8 @@ def batch_to_device(
             dataloader_idx: The index of the dataloader to which the batch belongs.
         """
         model = self.lightning_module
-
-        if model is not None:
+        if model is not None and not isinstance(self.training_type_plugin, DataParallelPlugin):
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:177]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — COMMENTED** [review]
LGTM!
